mlflow/xgboost.py

```diff
@@ -365,12 +374,39 @@ def record_eval_results(eval_results, metrics_logger):
             Create a callback function that records evaluation results.
             """
 
-            @exception_safe_function
-            def callback(env):
-                metrics_logger.record_metrics(dict(env.evaluation_result_list), env.iteration)
-                eval_results.append(dict(env.evaluation_result_list))
+            if LooseVersion(xgboost.__version__) >= LooseVersion("1.3.0"):
+                # In xgboost >= 1.3.0, user-defined callbacks should inherit
+                # `xgboost.callback.TrainingCallback`:
+                # https://xgboost.readthedocs.io/en/latest/python/callbacks.html#defining-your-own-callback  # noqa
+
+                class Callback(
+                    xgboost.callback.TrainingCallback, metaclass=ExceptionSafeAbstractClass,
+                ):
+                    def after_iteration(self, model, epoch, evals_log):
+                        """
+                        Run after each iteration. Return True when training should stop.
+                        """
+                        evaluation_result_dict = {}
+                        for data_name, metric_dict in evals_log.items():
+                            for metric_name, metric_values in metric_dict.items():
+                                key = "{}-{}".format(data_name, metric_name)
+                                evaluation_result_dict[key] = metric_values[-1]
+
+                        metrics_logger.record_metrics(evaluation_result_dict, epoch)
+                        eval_results.append(evaluation_result_dict)
+
+                        return False
```

Comment:
Why do we return `False`?

Reply:
It's just for indicating training should not stop:

https://github.com/dmlc/xgboost/blob/60cfd14349a3ed4593de276525fba068adcaea53/demo/guide-python/callbacks.py#L50

I'll insert a comment here for clarification.

Reply:
Returning nothing (= returning `None`) actually works too because `None` is falsey.